pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumerCoordinator.java

```diff
@@ -0,0 +1,302 @@
+      // sequence segment of the partition Group.
+      return;
+    }
+    String previousSegment = previousLLCSegmentName.getSegmentName();
+
+    SegmentDataManager segmentDataManager = _realtimeTableDataManager.acquireSegment(previousSegment);
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      _lock.lock();
+      try {
+        while (segmentDataManager == null) {
+          // if segmentDataManager == null, it means segment is not loaded in the server.
+          // wait until it's loaded.
+          while (!_condition.await(WAIT_INTERVAL_MS, TimeUnit.MILLISECONDS)) {
+            LOGGER.warn("Semaphore access denied to segment: {}. Waiting on previous segment: {} since: {} ms.",
+                currSegment.getSegmentName(), previousSegment, System.currentTimeMillis() - startTimeMs);
+            // waited until timeout, fetch previous segment again from ideal state as previous segment might be
+            // changed in ideal state.
+            previousLLCSegmentName = getPreviousSegment(currSegment);
+            if (previousLLCSegmentName == null) {
+              return;
+            }
+            previousSegment = previousLLCSegmentName.getSegmentName();
+          }
+          segmentDataManager = _realtimeTableDataManager.acquireSegment(previousSegment);
+        }
+      } finally {
+        _lock.unlock();
+      }
+    } finally {
+      if (segmentDataManager != null) {
+        _realtimeTableDataManager.releaseSegment(segmentDataManager);
+      }
+    }
+  }
+
+  /***
+   * @param currSegment is the segment of current helix transition.
+   * @param timeoutMs is max time to wait in millis
+   * @return true if previous Segment was registered to the server, else false.
+   * @throws InterruptedException
+   */
+  @VisibleForTesting
+  boolean awaitForPreviousSegmentSequenceNumber(LLCSegmentName currSegment, long timeoutMs)
+      throws InterruptedException {
+    long startTimeMs = System.currentTimeMillis();
+    int prevSeqNum = currSegment.getSequenceNumber() - 1;
+    _lock.lock();
+    try {
+      while (_maxSegmentSeqNumRegistered < prevSeqNum) {
+        // it means all segments until _maxSegmentSeqNumRegistered is not loaded in the server. Wait until it's loaded.
+        if (!_condition.await(timeoutMs, TimeUnit.MILLISECONDS)) {
+          LOGGER.warn("Semaphore access denied to segment: {}."
+                  + " Waiting on previous segment with sequence number: {} since: {} ms.", currSegment.getSegmentName(),
+              prevSeqNum, System.currentTimeMillis() - startTimeMs);
+          // waited until the timeout. Rely on ideal state now.
+          return false;
+        }
+      }
+
+      return (_maxSegmentSeqNumRegistered >= prevSeqNum);
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  @VisibleForTesting
+  @Nullable
+  LLCSegmentName getPreviousSegment(LLCSegmentName currSegment) {
```

Comment:
(minor) Make it more specific
```suggestion
  LLCSegmentName getPreviousSegmentFromIdealState(LLCSegmentName currSegment) {
```